Add Python SDK

danielvegamyhre opened the pull request:
Fixes #163 

These changes include the OpenAPI Go schema, which is used to generate the swagger.json file, which in turn is used to generate the Python SDK.

tenzen-y commented on api/jobset/v1alpha1/jobset_types.go:
```diff
@@ -89,6 +90,7 @@ type ReplicatedJobStatus struct {
 }
 
 // +genclient
+// +k8s:openapi-gen=true
```
@danielvegamyhre IIUC, we must add this marker to doc.go instead of adding this marker to types.

https://github.com/kubernetes/code-generator/blob/e60f5637f0bf9de7253d9258b4cb0b8958e2c5a1/examples/apiserver/apis/example/v1/doc.go#L17

danielvegamyhre replied:
that worked, thanks!

tenzen-y replied:
👍

tenzen-y commented on api/jobset/v1alpha1/jobset_types.go:
```diff
@@ -11,6 +11,7 @@ See the License for the specific language governing permissions and
 limitations under the License.
 */
 
+// +k8s:openapi-gen=true
```
Removing unused markers would be good.

tenzen-y commented on api/jobset/v1alpha1/jobset_types.go:
```diff
@@ -89,6 +90,7 @@ type ReplicatedJobStatus struct {
 }
 
 // +genclient
+// +k8s:openapi-gen=true
```
Same here.

ahg-g commented on client-go/applyconfiguration/internal/internal.go:
```diff
@@ -0,0 +1,58 @@
+/*
```
any idea what is this applyconfiguration pkg about?

danielvegamyhre replied:
looks like a way of type-safe doing server-side applies https://pkg.go.dev/k8s.io/client-go/applyconfigurations

ahg-g commented on hack/python-sdk/gen-sdk.sh:
```diff
@@ -0,0 +1,51 @@
+#!/usr/bin/env bash
+
+# Copyright 2021 The Kubeflow Authors.
```
add a reference to the original script

danielvegamyhre replied:
done

ahg-g commented on hack/python-sdk/gen-sdk.sh:
```diff
@@ -0,0 +1,51 @@
+#!/usr/bin/env bash
```
when should this script be run, should we run it as part of `make generate`?

danielvegamyhre replied:
yep i think we should, added

danielvegamyhre commented:
Loks like we need to download a java runtime in the CI script so we can run the JAR used in gen-sdk.sh

tenzen-y commented on sdk/python/.gitlab-ci.yml:
Why do we need this config for GitLab CI?

danielvegamyhre replied:
it is autogenerated

tenzen-y commented on sdk/python/.travis.yml:
Why do we need this config for Travis CI?

danielvegamyhre replied:
it is autogenerated

tenzen-y commented on sdk/python/git_push.sh:
Why do we need this script?

danielvegamyhre replied:
it is autogenerated

tenzen-y commented:
@danielvegamyhre Also, adding examples using our PyThon SDK with Dynamic Client would be good. And then running the examples in CI every CI would be great.

But we can work on that in follow-up PRs.

ahg-g commented on hack/python-sdk/post_gen.py:
```diff
@@ -0,0 +1,59 @@
+#!/usr/bin/env python
+
+# Copyright 2021 The Kubeflow Authors.
+#
```
ditto, please add a reference

ahg-g commented on hack/python-sdk/swagger.json:
```diff
@@ -0,0 +1,222 @@
+{
```
what generates this file and any idea what purpose it serves?

danielvegamyhre replied:
It is generated by `hack/swagger/main.go`. It serves as input to openapi-generator-cli.jar which generates the Python SDK.

ahg-g commented on hack/swagger/main.go:
```diff
@@ -0,0 +1,83 @@
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+package main
```
where/when does this program get executed?

danielvegamyhre replied:
It's executed on line 45 of `gen-sdk.sh`

ahg-g commented on sdk/python/.gitignore:
```diff
@@ -0,0 +1,66 @@
+# Byte-compiled / optimized / DLL files
```
was this file auto-generated?

danielvegamyhre replied:
Yes

ahg-g commented:
> Update: after adding `post_gen.py` to fix issues in the python sdk after generating it, the newest error when running the python sdk unit tests is:
> 
> ```
> jobset/models/jobset_v1alpha1_replicated_job.py:66: in __init__
>     self.template = template
> _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
> 
> self = {'name': '0',
>  'network': {'enable_dns_hostnames': True},
>  'replicas': 56,
>  'template': None}, template = None
> 
>     @template.setter
>     def template(self, template):
>         """Sets the template of this JobsetV1alpha1ReplicatedJob.
>     
>     
>         :param template: The template of this JobsetV1alpha1ReplicatedJob.  # noqa: E501
>         :type: K8sIoApiBatchV1JobTemplateSpec
>         """
>         if self.local_vars_configuration.client_side_validation and template is None:  # noqa: E501
> >           raise ValueError("Invalid value for `template`, must not be `None`")  # noqa: E501
> E           ValueError: Invalid value for `template`, must not be `None`
> 
> jobset/models/jobset_v1alpha1_replicated_job.py:156: ValueError
> ===================================================================== short test summary info =====================================================================
> FAILED test/test_jobset_v1alpha1_replicated_job.py::TestJobsetV1alpha1ReplicatedJob::testJobsetV1alpha1ReplicatedJob - ValueError: Invalid value for `template`, must not be `None`
> FAILED test/test_jobset_v1alpha1_job_set_spec.py::TestJobsetV1alpha1JobSetSpec::testJobsetV1alpha1JobSetSpec - ValueError: Invalid value for `template`, must not be `None`
> FAILED test/test_jobset_v1alpha1_job_set_list.py::TestJobsetV1alpha1JobSetList::testJobsetV1alpha1JobSetList - ValueError: Invalid value for `template`, must not be `None`
> FAILED test/test_jobset_v1alpha1_job_set.py::TestJobsetV1alpha1JobSet::testJobsetV1alpha1JobSet - ValueError: Invalid value for `template`, must not be `None`
> =================================================================== 4 failed, 5 passed in 0.28s ===================================================================
> ```

Is this new? if so, any idea why this error now started to trigger?

danielvegamyhre commented:
@ahg-g No it isn't new, it's just the only remaining error currently. Sorry I should have phrased the previous comment as "remaining error" rather than "new error."

danielvegamyhre commented:
Here is the issue summary, @tenzen-y do you have any idea what could cause a unit test file to set a required parameter as None?

swagger.json correctly defines `replicatedJob.template` as required: https://github.com/kubernetes-sigs/jobset/pull/175/files#diff-ebfb097c36af64a643252d0451d2db348d52d0c9826dc95f23474fec69debb8eR151

replicatedJob constructor itself seems to allow setting template as None: https://github.com/kubernetes-sigs/jobset/pull/175/files#diff-08ce0660da8f10cd2049e709cd704142b5f44cf675931ef2005b217a095454b0R66

However, replicatedJob.template setter method has validation preventing setting template as None if client side validation is enabled:
https://github.com/kubernetes-sigs/jobset/pull/175/files#diff-08ce0660da8f10cd2049e709cd704142b5f44cf675931ef2005b217a095454b0R148-R156

Yet test file creates instance with empty template: https://github.com/kubernetes-sigs/jobset/pull/175/files#diff-638109b6b304ab834aca870d16a1a6fa347f2b1e242a79f0debb455702b3ccafR43-R49

tenzen-y commented:
Let me check.

charles-chenzz commented:
have we consider separate it into multiple PRs? eg: auto generated file in a PR, sdk PR etc... which I think would be easier to track issues/bugs

charles-chenzz commented:
> IIUC, almost files are automatically generated (including SDK, uni test, etc...).

ah I see then, (haven't review the file yet, sorry for adding noise)

danielvegamyhre commented:
@tenzen-y @ahg-g I finally figured it out, it seems obvious in hindsight.

The code generator lacks the ability to infer arbitrary fully-qualified Python package names and import them (e.g. determine that if the type is `V1JobTemplateSpec` in `swagger.json`, it cannot determine that it in the test file it must do something like `from kubernetes.client.models.v1_job_template_spec import V1JobTemplateSpec` so that it can be used for the `template=V1JobTemplateSpec(...)` param of the replicatedJob.  

Since the code generator doesn't know what a `V1JobTemplateSpec` is, it just sets the template to None in the test files without throwing any errors or logging a warning (which I wish it did, as debugging this would have been easier).

I updated the `post_gen.py` script to include the Python `kubernetes` package in the `test-requirements.txt` file and modify the test_files to do `from kubernetes.client.models.v1_job_template_spec import V1JobTemplateSpec` and replace `template = None` with `template = V1JobTemplateSpec()`, and now the Python SDK unit tests are passing 🎉

danielvegamyhre commented:
@ahg-g to make reviewing easier, I've grouped the files into manually written vs. auto-generated:

Manually written:
- Makefile
- everything in `api/jobset/v1alpha2/` except `openapi_generated.go`
- everything in `hack/python-sdk`
- everything in `hack/swagger`
- `go.mod` (manually tweaked package versions to resolve dependency conflicts)

Auto-generated:
- `api/jobset/v1alpha2/openapi_generated.go`
- everything in `client-go/`
- everything in `sdk/python`

danielvegamyhre commented:
For some reason the `gen-sdk.sh` script works locally but in CI it crashes on the first line that isn't just an `echo` command... need to investigate this further

danielvegamyhre commented:
@ahg-g I managed to fix the bash script bug and all of the tests (including Python SDK unit tests) are now passing, so this is ready for review

ahg-g commented:
@danielvegamyhre this is great, what an incredible effort!

A couple of comments:
1) There is a binary file named `sdk/python/.coverage`, do we need it?
2) We added the subdomain field to the Network struct, I wonder if that should have triggered a failure to the verify-main test since I don't see it in your PR yet?

danielvegamyhre commented:
> 1. There is a binary file named `sdk/python/.coverage`, do we need it?
No, it just contains info about what lines have code have executed for generating test coverage reports. It should have been ignored by the .gitignore but I accidentally added it before updating the .gitignore - thanks for catching this, I've removed it.

> 2. We added the subdomain field to the Network struct, I wonder if that should have triggered a failure to the verify-main test since I don't see it in your PR yet?
Ah, I rebased after #171 was merged but forgot to force-push so it didn't upload. Just force-pushed now. 

I also updated the bash scripts to only install Java and Python if they aren't installed already (basically just evaluating if `which java` or `which python3` returns empty or not - perhaps there's a better way to do this but haven't found one yet).

tenzen-y commented on hack/python-sdk/test-sdk.sh:
```diff
@@ -0,0 +1,36 @@
+fi
+
+if [ -z `which pip` ]; then
+    apt update -y
+    apt-get install python3-pip -y
```
Avoiding depending on the running env would be good. But we can work on it in follow-ups.

danielvegamyhre replied:
Makes sense. In a follow up PR I will try to download and build from source, rather than use an environment specific package manager.

tenzen-y replied:
Sounds good.

tenzen-y reviewed (commented):
Thanks for the great work!

ahg-g commented:
/label tide/merge-method-squash

tenzen-y commented:
LGTM
/assign @ahg-g

danielvegamyhre commented:
Also created #189 to add example using python sdk + dynamic client and put it in the "nice to have" category for v0.2.0 release.

tenzen-y commented:
> sdk + dynamic client and put it in the "nice to have" category for

@danielvegamyhre Thanks!

ahg-g commented:
/lgtm

This is great!
